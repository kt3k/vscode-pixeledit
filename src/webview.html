<!DOCTYPE html>
<html>
  <head>
    <link rel="stylesheet" href="${styleUri}" />
    <link rel="stylesheet" href="${tailwindUri}" />
  </head>
  <body>
    <div class="board">
      <div class="mini-canvas-wrapper hidden">
        <span>
          <canvas class="mini-canvas" id="canvas-mini"></canvas>
        </span>
      </div>
      <canvas
        id="canvas"
        style="background: linear-gradient(#303, #202)"
      ></canvas>
      <div class="js-tools fixed left-0 grid grid-cols-2">
        <span
          class="item border border-gray-300 w-12 h-12 flex justify-center items-center text-2xl"
          data-index="0"
          data-tool="pen"
        >✏️</span>
        <span
          class="item border border-gray-300 w-12 h-12 flex justify-center items-center text-2xl"
          data-index="1"
          data-tool="eraser"
        >消</span>
        <span
          class="item border border-gray-300 w-12 h-12 flex justify-center items-center text-2xl"
          data-index="2"
          data-tool="fill"
        >塗</span>
      </div>
      <div id="palette" class="js-palette"></div>
    </div>
  </body>
  <script src="${scriptUri}" type="module"></script>
</html>
