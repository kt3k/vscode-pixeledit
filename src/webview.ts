// Copyright 2022-2024 Yoshiya Hinosawa. All rights reserved. MIT license.
// Copyright 2021 PixelCraft. All rights reserved. MIT license.

/// <reference lib="dom" />
/// <reference path="../node_modules/@types/vscode-webview/index.d.ts" />

import { type Context, register, Signal } from "@kt3k/cell"

import type {
  Color,
  Edit,
  ExtensionMessageEvent,
  WebviewMessage,
} from "./types.ts"

let board: Board

type Tool = "pen" | "eraser" | "fill"
const currentTool = new Signal<Tool>("pen")
const paletteColors = new Signal<Color[]>([
  [0, 0, 0, 255],
  [0x7c, 0x7c, 0x7c, 255],
  [0xbc, 0xbc, 0xbc, 255],
  [0xf8, 0xf8, 0xf8, 255],
  [0xfc, 0xfc, 0xfc, 255],
  [237, 28, 36, 255],
  [255, 127, 39, 255],
  [255, 242, 0, 255],
  [34, 177, 36, 255],
  [0, 162, 232, 255],
  [63, 72, 204, 255],
  [163, 73, 164, 255],
  [255, 255, 255, 255],
  [195, 195, 195, 255],
  [185, 122, 87, 255],
  [255, 174, 201, 255],
  [255, 201, 14, 255],
  [239, 228, 176, 255],
  [181, 230, 29, 255],
  [153, 217, 234, 255],
  [112, 146, 190, 255],
  [200, 191, 231, 255],
])
const currentColor = new Signal<Color>([0, 0, 0, 255])

function toCssColor(c: Color) {
  return `rgba(${c[0]},${c[1]},${c[2]},${c[3] / 255})`
}

function toHex(c: Color) {
  const r = c[0].toString(16).padStart(2, "0")
  const g = c[1].toString(16).padStart(2, "0")
  const b = c[2].toString(16).padStart(2, "0")
  return "#" + r + g + b
}

function saveEdit(stroke: [number, number][], color: Color) {
  postMessage({
    type: "edit",
    edit: {
      color,
      stroke,
    },
  })
}

class Board {
  /** The canvas */
  canvas: HTMLCanvasElement
  /** The canvas */
  miniCanvas: HTMLCanvasElement
  /** The canvas context */
  ctx: CanvasRenderingContext2D
  /** The canvas context */
  miniCtx: CanvasRenderingContext2D
  /** Image data width */
  dataWidth: number
  /** Image data height */
  dataHeight: number
  /** Canvas element width */
  canvasWidth: number
  /** Canvas element height */
  canvasHeight: number
  /** pixel data array */
  data: Color[][]
  color: Color = [0, 0, 0, 0]
  constructor(dataWidth: number, dataHeight: number) {
    document.querySelector<HTMLElement>(".mini-canvas-wrapper")!.classList
      .toggle("hidden", false)
    this.canvas = document.querySelector("#canvas")!
    this.miniCanvas = document.querySelector("#canvas-mini")!
    this.canvas.width = 10 * dataWidth
    this.canvas.height = 10 * dataHeight
    this.miniCanvas.width = dataWidth
    this.miniCanvas.height = dataHeight
    this.dataWidth = dataWidth
    this.dataHeight = dataHeight
    this.canvas.style.display = "block"
    this.canvas.style.height =
      Math.floor((dataHeight / dataWidth) * this.canvas.clientWidth) + "px"
    this.canvasWidth = +this.canvas.width
    this.canvasHeight = +this.canvas.height
    this.ctx = this.canvas.getContext("2d")!
    this.miniCtx = this.miniCanvas.getContext("2d")!
    this.ctx.fillStyle = "rgba(255,255,255,0)"
    this.ctx.fillRect(0, 0, this.canvasWidth, this.canvasHeight)
    this.miniCtx.fillStyle = "rgba(255,255,255,0)"
    this.ctx.fillRect(0, 0, this.canvasWidth, this.canvasHeight)
    this.data = [...Array(this.dataWidth)].map((_e) =>
      Array(this.dataHeight).fill([255, 255, 255, 0])
    )

    this.canvas.addEventListener("mouseup", (e) => {
      const rect = this.canvas.getBoundingClientRect()
      let x = e.clientX - rect.left
      let y = e.clientY - rect.top
      x = Math.floor(this.dataWidth * x / this.canvas.clientWidth)
      y = Math.floor(this.dataHeight * y / this.canvas.clientHeight)
      if (!this.validCoords(x, y)) {
        return
      }
      if (currentTool.get() === "fill") {
        filler(x, y, this.data[x][y])
      } else if (currentTool.get() === "eraser") {
        this.draw(x, y, [0, 0, 0, 0])
        saveEdit([[x, y]], [0, 0, 0, 0])
      } else {
        // Pen tool
        this.draw(x, y, currentColor.get())
        saveEdit([[x, y]], currentColor.get())
      }
    })
  }

  validCoords(x: number, y: number) {
    return x >= 0 && x < this.dataWidth && y >= 0 && y < this.dataHeight
  }

  draw(x: number, y: number, color: Color) {
    this.data[x][y] = color
    this.ctx.clearRect(
      Math.floor(x * (this.canvasWidth / this.dataWidth)),
      Math.floor(y * (this.canvasHeight / this.dataHeight)),
      Math.floor(this.canvasWidth / this.dataWidth),
      Math.floor(this.canvasHeight / this.dataHeight),
    )
    this.ctx.fillStyle = toCssColor(color)
    this.ctx.fillRect(
      Math.floor(x * (this.canvasWidth / this.dataWidth)),
      Math.floor(y * (this.canvasHeight / this.dataHeight)),
      Math.floor(this.canvasWidth / this.dataWidth),
      Math.floor(this.canvasHeight / this.dataHeight),
    )
    this.miniCtx.clearRect(x, y, 1, 1)
    this.miniCtx.fillStyle = toCssColor(color)
    this.miniCtx.fillRect(x, y, 1, 1)
  }

  async update(dataUri: string, edits: Edit[]) {
    this.ctx.clearRect(0, 0, this.canvasWidth, this.canvasHeight)
    await this.importImage(dataUri)
    for (const edit of edits) {
      this.applyEdit(edit)
    }
  }

  applyEdit(edit: Edit) {
    for (const [x, y] of edit.stroke) {
      this.draw(x, y, edit.color)
    }
  }

  importImage(uri: string): Promise<void> {
    const uimg = new Image()
    uimg.src = uri
    uimg.width = this.dataWidth
    uimg.height = this.dataHeight
    return new Promise<void>((resolve, reject) => {
      uimg.onload = () => {
        const pxc = document.createElement("canvas")
        pxc.width = this.dataWidth
        pxc.height = this.dataHeight
        const pxctx = pxc.getContext("2d")!
        pxctx.drawImage(uimg, 0, 0, this.dataWidth, this.dataHeight)
        for (let i = 0; i < this.dataWidth; i++) {
          for (let j = 0; j < this.dataHeight; j++) {
            const pixel = pxctx.getImageData(i, j, 1, 1).data
            this.draw(i, j, [pixel[0], pixel[1], pixel[2], pixel[3]])
          }
        }
        resolve()
      }
      uimg.onerror = (e) => reject(e)
    })
  }

  static async import(uri: string, edits: Edit[]): Promise<Board> {
    const img = await loadImage(uri)
    const board = new Board(img.width, img.height)
    await board.importImage(uri)
    for (const edit of edits) {
      board.applyEdit(edit)
    }
    return board
  }

  exportImage() {
    const canvas = document.createElement("canvas")
    canvas.width = this.dataWidth
    canvas.height = this.dataHeight
    const ctx = canvas.getContext("2d")!
    this.data.forEach((row, i) => {
      row.forEach((color, j) => {
        ctx.fillStyle = toCssColor(color)
        ctx.fillRect(i, j, 1, 1)
      })
    })
    return canvas.toDataURL("image/png")
  }
}

function filler(x: number, y: number, cc: Color) {
  if (x >= 0 && x < board.dataWidth && y >= 0 && y < board.dataHeight) {
    if (
      JSON.stringify(board.data[x][y]) == JSON.stringify(cc) &&
      JSON.stringify(board.data[x][y]) != JSON.stringify(currentColor.get())
    ) {
      board.draw(x, y, currentColor.get())
      filler(x + 1, y, cc)
      filler(x, y + 1, cc)
      filler(x - 1, y, cc)
      filler(x, y - 1, cc)
    }
  }
}

function loadImage(uri: string): Promise<HTMLImageElement> {
  return new Promise<HTMLImageElement>((resolve, reject) => {
    const img = new Image()
    img.src = uri
    img.onload = () => resolve(img)
    img.onerror = (e) => reject(e)
  })
}

/** Tools UI component */
function Tools({ on, queryAll }: Context) {
  on("click", ".item", ({ target }) => {
    const item = target as HTMLElement
    // deno-lint-ignore no-explicit-any
    currentTool.update(item.dataset.tool as any)
    queryAll<HTMLElement>(".item").forEach((x) => {
      x.classList.toggle("bg-gray-500", false)
    })
    item.classList.toggle("bg-gray-500", true)
  })
}

/** Palette UI Component */
function Palette({ el, on, queryAll, subscribe }: Context) {
  on("click", ".item", (e) => {
    const item = e.target as HTMLElement
    currentColor.update(JSON.parse(item.dataset.color!))
  })

  subscribe(paletteColors, (colors) => {
    el.innerHTML = colors.map((color) =>
      `<span
         class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer"
         data-color="${JSON.stringify(color)}"
         style="background-color: ${toCssColor(color)};"
      >
        <span
          class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center"
          style="font-size: 0.5rem;"
        >
          ${toHex(color)}
        </span>
      </span>`
    ).join("\n")
  })

  subscribe(currentColor, (color) => {
    queryAll<HTMLElement>(".item").forEach((x) => {
      if (x.dataset.color === JSON.stringify(color)) {
        x.style.boxShadow = "0px 0px 1px 1px white inset"
      } else {
        x.style.boxShadow = ""
      }
    })
  })
}

const vscode = acquireVsCodeApi()
function postMessage(message: WebviewMessage) {
  vscode.postMessage(message)
}

globalThis.addEventListener("message", async (e: ExtensionMessageEvent) => {
  console.log("extension -> webview " + e.data.type, e)
  switch (e.data.type) {
    case "init": {
      board = await Board.import(e.data.dataUri, e.data.edits)
      break
    }
    case "new": {
      // TODO: Implement new
      break
    }
    case "getBytes": {
      postMessage({
        type: "response",
        requestId: e.data.requestId,
        body: board.exportImage(),
      })
      break
    }
    case "update": {
      board.update(e.data.doc.dataUri, e.data.doc.edits)
      break
    }
  }
})

register(Palette, "js-palette")
register(Tools, "js-tools")
postMessage({ type: "ready" })
