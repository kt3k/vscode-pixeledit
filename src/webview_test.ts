// Copyright 2022-2024 Yoshiya Hinosawa. All rights reserved. MIT license.
/// <reference lib="deno.ns" />

import { assertEquals } from "@std/assert"
import { toCssColor, toHex } from "./webview.ts"
import { DOMParser } from "@b-fuze/deno-dom"
import { mount } from "@kt3k/cell"
import { assertSnapshot } from "@std/testing/snapshot"

const domParser = new DOMParser()

Deno.test("toHex", () => {
  assertEquals(toHex([0, 0, 0, 255]), "#000000")
})

Deno.test("toCssColor", () => {
  assertEquals(toCssColor([0, 0, 0, 255]), "rgba(0,0,0,1)")
})

Deno.test("js-tools", async (t) => {
  // deno-lint-ignore no-explicit-any
  ;(globalThis as any).document = domParser.parseFromString(
    `<div class="js-tools"></div>`,
    "text/html",
  )
  mount()
  await assertSnapshot(t, document.querySelector(".js-tools")?.innerHTML)
})

Deno.test("js-palette", async (t) => {
  // deno-lint-ignore no-explicit-any
  ;(globalThis as any).document = domParser.parseFromString(
    `<div class="js-palette"></div>`,
    "text/html",
  )
  mount()
  await assertSnapshot(t, document.querySelector(".js-palette")?.innerHTML)
})
