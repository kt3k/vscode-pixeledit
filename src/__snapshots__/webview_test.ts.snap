export const snapshot = {};

snapshot[`js-tools 1`] = `
'
    <span class="item border border-gray-300 w-12 h-12 flex justify-center items-center text-2xl bg-gray-500" data-tool="pen">✒️</span>
  
    <span class="item border border-gray-300 w-12 h-12 flex justify-center items-center text-2xl" data-tool="eraser">消</span>
  
    <span class="item border border-gray-300 w-12 h-12 flex justify-center items-center text-2xl" data-tool="fill">塗</span>
  '
`;

snapshot[`js-palette 1`] = `
'<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[0,0,0,255]" style="background-color: rgba(0,0,0,1);">
        <span class="text-white bg-opacity-70 absolute bottom-0 w-full text-center bg-yellow-600" style="font-size: 0.5rem;">
          #000000
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[124,124,124,255]" style="background-color: rgba(124,124,124,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #7c7c7c
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[188,188,188,255]" style="background-color: rgba(188,188,188,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #bcbcbc
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[248,248,248,255]" style="background-color: rgba(248,248,248,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #f8f8f8
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[252,252,252,255]" style="background-color: rgba(252,252,252,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #fcfcfc
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[237,28,36,255]" style="background-color: rgba(237,28,36,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #ed1c24
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[255,127,39,255]" style="background-color: rgba(255,127,39,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #ff7f27
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[255,242,0,255]" style="background-color: rgba(255,242,0,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #fff200
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[34,177,36,255]" style="background-color: rgba(34,177,36,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #22b124
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[0,162,232,255]" style="background-color: rgba(0,162,232,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #00a2e8
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[63,72,204,255]" style="background-color: rgba(63,72,204,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #3f48cc
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[163,73,164,255]" style="background-color: rgba(163,73,164,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #a349a4
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[255,255,255,255]" style="background-color: rgba(255,255,255,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #ffffff
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[195,195,195,255]" style="background-color: rgba(195,195,195,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #c3c3c3
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[185,122,87,255]" style="background-color: rgba(185,122,87,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #b97a57
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[255,174,201,255]" style="background-color: rgba(255,174,201,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #ffaec9
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[255,201,14,255]" style="background-color: rgba(255,201,14,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #ffc90e
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[239,228,176,255]" style="background-color: rgba(239,228,176,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #efe4b0
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[181,230,29,255]" style="background-color: rgba(181,230,29,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #b5e61d
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[153,217,234,255]" style="background-color: rgba(153,217,234,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #99d9ea
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[112,146,190,255]" style="background-color: rgba(112,146,190,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #7092be
        </span>
      </span>
<span class="item relative flex items-end w-12 h-12 border border-transparent hover:border-gray-500 cursor-pointer" data-color="[200,191,231,255]" style="background-color: rgba(200,191,231,1);">
        <span class="text-white bg-gray-800 bg-opacity-70 absolute bottom-0 w-full text-center" style="font-size: 0.5rem;">
          #c8bfe7
        </span>
      </span>'
`;
